pkg/cloudprovider/rhsm/subscription_manager.go

```diff
@@ -69,40 +64,40 @@ func NewRedHatSubscriptionManager(offlineToken string) (RedHatSubscriptionManage
 	if offlineToken == "" {
 		return nil, errors.New("RedHatSubscriptionManager offline token cannot be empty")
 	}
-
 	return &defaultRedHatSubscriptionManager{
-		client: &http.Client{
-			Timeout: 10 * time.Second,
-		},
-		authURL:         "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token",
-		apiURL:          "https://api.access.redhat.com/management/v1/systems",
-		offlineToken:    offlineToken,
-		requestsLimiter: 100,
+		apiURL: "https://api.access.redhat.com/management/v1/systems",
+		client: newOAuthClientWithRefreshToken(offlineToken, "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"),
 	}, nil
 }
 
-func (d *defaultRedHatSubscriptionManager) UnregisterInstance(machineName string) error {
-	if d.credentials == nil {
-		klog.Info("access token has been expired or not initialized, refreshing token")
-		if err := d.refreshToken(); err != nil {
-			return fmt.Errorf("failed to refresh offline token: %v", err)
-		}
+func newOAuthClientWithRefreshToken(refreshToken string, tokenURL string) *http.Client {
+	ctx := context.Background()
+	// Use the custom HTTP client when requesting a token.
+	httpClient := &http.Client{Timeout: 5 * time.Second}
```

Comment:
By better looking into it this timeout seem to apply for the token request only, I should add a timeout for the API calls

Reply:
Since you are only adding a timeout to the client nothing else, you can set the timeout explicitly on the client instead of passing it by the context which might be overwritten by the new context.

Reply:
```go
conf := &oauth2.Config{
		ClientID: "rhsm-api",
		Endpoint: oauth2.Endpoint{
			TokenURL: tokenURL,
		},
	}
client := conf.Client(ctx, tok)
client.Timeout =  5 * time.Second
```

Reply:
btw, why did you change the timeout from 10 to 5?

Reply:
> Since you are only adding a timeout to the client nothing else, you can set the timeout explicitly on the client instead of passing it by the context which might be overwritten by the new context.

The client that I'm passing through the context is used by the retrieve token logic (see https://godoc.org/golang.org/x/oauth2#example-Config--CustomHTTP), if I don't provide it the default client is used instead. 
So it's ok to pass it but it's not enough. I should set the default timeout to the returned client as well as you suggested.